### ISSUE TITLE
Use HTTP 410 and an according text on the URL of a deleted toot

### ISSUE THREAD

**Issue by allo-** [tootsuite/mastodon#10640]
### Feature Request
Indicate deleted toots by HTTP Error 410 and a according status text like "The linked toot was deleted", when someone tries to load them using a link.

### Current Behaviour

When you visit a link to a deleted toot, it wrongly states "something went wrong on our side", suggesting that visiting it later may help. This is wrong for 404, but even more wrong for deleted toots, which should have status 410 (Gone) in the best case.

I had a deleted toot in my timeline. When I visited the page on the original instance, I got an error page, that stated the fault is with the remote instance.

### Suggested changes:

- Keep a record of toot IDs that existed
- When a user tries to load a toot that existed but is deleted, show a "This toot was deleted" message
- Use HTTP status 410 to indicate the deletion as a machine readable status.

### Further ideas

In addition, remote instances could use the status to delete the toot from their own databases, when they try to load it (again), and did not get the delete message before.  
Furthermore they may decide to show a "toot was deleted" message instead of silently removing it.

**Comment by ThibG** [comment]
We may want to provide some sort of deniability for deleted toots, so I'm not sure we should actually use 410 for deleted toots.

I think we should delete toots and accounts when they raise a 410, though

**Comment by allo-** [comment]
I am not sure about the deniability.

1) When there is no web archive, then there is not much information, except that there was a toot.
2) When the toot was linked somewhere or belongs to a thread that references its metadata, the existence is proven anyway (except for doctored metadata, but in the average case it is highly likely that there was a toot)
3) A link from a website resulting in 404 instead of 410 may help to deny that a made up quote was tooted.

On 2: When there is the metadata, 404 vs. 410 only documents if it was deleted or if there is a technical problem (maybe data loss on the instance and the toot was not in the backup, yet).
On 3: there is deniability for this specific toot anyway, as some made up quote could link to a toot deleted long time ago, that did not have anything to do with the discussion. Not sure about monotonic toot IDs, though.

A simple implementation before deciding on the deniability considerations would be to change the error message on toot URLs from "something is wrong on our side" to "This toot does not exist", for *any* non-existing toot ID, disregarding if it ever existed or not.

Personally I would associate "something is wrong on our side" with 5xx error codes and not 4xx. 404 usually happens because somebody has a typo in the URL, copied an incomplete URL or links something deleted (that may or may not result in 410 instead).

**Comment by hook321** [comment]
I just tested this, and it appears to return status 404 for deleted posts, which isn't good either.

Due to the fact that people oftentimes use HTTP status codes incorrectly, I think it's a bad idea to delete posts when status 410 is returned. There might also be some cases where an instance owner might want to shut down their instance but not request that the content get deleted on other instances, and then might later let their domain expire since they potentially don't have a need for it anymore. If instances delete things locally when status 410 is returned then someone could easily buy the domain, have the server return status 410 for all requests, and effectively delete posts that came from that instance, for malicious purposes.

**Comment by allo-** [comment]
I disagree that posts from closed instances should be retained, but deletion based on HTTP status and DNS failures is not that easy either.

I disagree, because the Ex-Users will not have *any* control about their status anymore, so it would be reasonable to delete them on other instances, such that there aren't status around, that cannot be deleted by their owners. Furthermore, some ex-users may react to the owner closing the instance with "I wanted to delete my account anyway, now I can just stay inactive and it will be gone.

In addition there is a risk, that the new domain owner starts reusing the domain for an instance and that old status are associated with new users on other instances but not on his own. So his own users would not see anything weird, but the rest of the network will have an inconsistent state and associate wrong posts with the new accounts.

A malicious new owner could probably do quite a bit more nasty stuff. Deleting old status is one the least problem, when someone who bought the domain wants to mess with the old data on other instances by abusing the domain. And if possible, it should be the only thing he can do, but to make this sure the status would need to be signed.

This is not easily possible with HTTP 404 or DNS failures, because you cannot identify if the instance is closed or if there are just a technical problem. You do not want to delete all status, just because the domain owner did not setup the DNS correctly when moving servers.

In this regard, serving status 410 would be a great option. If you close your account and the deletion message did not reach all instances, instances accessing your profile can delete the remaining status. When the admin decides to close his instance, he can serve 410 on all URLs for some time, so other instances can see that they should delete everything from this instance.

In addtion, a "server is closed" message would be nice (tell me, when it already exists). The old admin could send every few weeks until the domain expires "server closed" to all known instances, that did not acknowledge it, yet. Others may see the status 410 on some endpoint that should never serve it (maybe something like the global timeline) and then delete the status from the instance.

> Due to the fact that people oftentimes use HTTP status codes incorrectly, I think it's a bad idea to delete posts when status 410 is returned
This would be a problem, but I never saw a 410 by accident. 404, 401, 403, 5xx may happen, but 410 should not be served for any technical failure. I do not know what software would send this when it is not desired.

**Comment by ClearlyClaire** [comment]
> Revealing a post ID as (possibly formerly) valid leaks the information that the account posted something at a given time (the post ID encodes the time the post was authored), which can be considered private information. If we do consider that it is ok to leak this information, it may make sense to return 410 for deleted posts as well.

Thinking more about this, a possible alternative would be to return a 410 for non-existing posts earlier than current time. This would change the issue to being able to prove the existence of a private post you know the URL of, as this would be an existing post returning 404 rather than 410. I'm not sure which is worse.

**Comment by allo-** [comment]
Both violate HTTP semantics and can cause problems. For example, search engines may assume that content that is gone will never return. Content that is (currently) not found may be revisited in the future. For private statuses, 403 would probably be the appropriate HTTP status code, telling clients, bots and browsers that they have no access to that information.
HTTP 401 would also be semantically correct, but it is strongly associated with HTTP auth, and for Mastodon the way to access the status would be to use an account that follows the user, which is more than just logging in.

**Comment by ClearlyClaire** [comment]
> Both violate HTTP semantics and can cause problems. For example, search engines may assume that content that is gone will never return.

We can return different HTTP statuses for ActivityPub requests and HTML requests, but yeah, the base issue remains the same.

And treating a 404 as a deletion is probably a bad idea, because of private posts, and because of the potential for temporary deployment failures.

> Content that is (currently) not found may be revisited in the future.

A post deleted in Mastodon currently remains deleted. However, this is something we want to leave the door open for, as mentioned in https://github.com/mastodon/mastodon/issues/17553

> For private statuses, 403 would probably be the appropriate HTTP status code, telling clients, bots and browsers that they have no access to that information.

That would still confirm existence of the post, which is something we are currently avoiding.

**Comment by Laxystem** [comment]
> That would still confirm existence of the post, which is something we are currently avoiding.

Frankly, I don't see the issue here. What security implications does knowing a post exists pose? You can find this out via ActivityPub anyway due to the nature of the protocol.
